---
title: "EWAS and variance explained"
output: html_notebook
---

The efficiency of EWAS, for finding positions in the genome associated with a complex trait for a given sample size, far outstrips the efficiency of GWAS. For example, The Wahl et al (2017) BMI EWAS found 187 EWAS hits, based on 450k DNAm sites that represents <2% of all DNAm sites, using 10k samples. By contrast, Locke et al (2015) identified only 97 GWAS hits using 330k samples. These 97 variants explain about 5% of the phenotypic variance of BMI. So if BMI has a heritability of say 50%, and 10% of the heritability is captured by the 100 largest variants, then the remaining variants must be composed than more than 10 times the 97 original variants. It will be highly polygenic, i.e 10000 causal variants is likely an underestimate, but let's use that as an exampl, then any one GWAS hit will explain on average $R^2 = 0.005%$ of the variance. To detect an effect of such magnitude you need a large sample size i.e.

$$
t^2 = \frac{R^2(n-1)}{1-R^2}
$$
So for a significance threshold of 5e-8 this is the sample size we'd need to detect an effect that has the mean $R^2$ value

```{r}
tval <- qnorm(5e-8, low=F)
n <- tval^2 * (1 - 0.00005) / 0.00005 - 1
n
```


## Wahl et al

These are the p-values for the significant EWAS hits from Wahl et al (2017)

```{r}
pvals <- c(3.20E-08, 1.30E-08, 7.20E-10, 1.20E-09, 1.20E-14, 6.00E-08, 4.60E-08, 6.20E-08, 1.40E-10, 2.30E-18, 8.10E-08, 3.00E-08, 3.80E-13, 2.10E-17, 1.40E-08, 3.20E-12, 2.90E-27, 4.40E-13, 2.20E-09, 2.50E-10, 1.60E-08, 1.80E-10, 4.00E-10, 3.10E-10, 4.70E-08, 3.70E-09, 7.70E-09, 5.80E-13, 3.20E-12, 6.00E-09, 4.70E-08, 1.50E-08, 1.00E-11, 5.10E-10, 7.20E-09, 5.70E-10, 2.10E-08, 1.50E-08, 1.40E-08, 5.50E-08, 4.40E-08, 4.00E-10, 3.40E-10, 7.70E-08, 2.90E-10, 4.00E-12, 6.70E-08, 3.10E-08, 5.30E-08, 2.70E-11, 4.90E-14, 1.50E-08, 8.10E-18, 6.00E-12, 6.40E-08, 8.50E-08, 3.00E-10, 2.90E-09, 9.10E-10, 3.00E-08, 2.90E-15, 1.30E-09, 1.20E-10, 2.50E-09, 4.90E-08, 8.20E-09, 1.10E-08, 3.40E-09, 1.60E-09, 1.60E-09, 3.70E-09, 3.00E-09, 8.20E-19, 9.60E-10, 2.40E-10, 2.30E-08, 6.20E-10, 2.60E-09, 5.90E-09, 7.00E-12, 1.40E-08, 1.50E-09, 5.30E-10, 3.20E-10, 2.30E-10, 9.40E-09, 3.60E-09, 1.40E-10, 1.60E-12, 9.00E-08, 1.10E-08, 4.20E-08, 7.80E-10, 1.60E-10, 3.00E-08, 8.80E-08, 4.50E-09, 1.10E-10, 4.30E-08, 1.90E-12, 9.20E-08, 3.90E-11, 1.40E-11, 6.50E-11, 5.00E-08, 6.10E-11, 5.80E-15, 4.50E-09, 3.30E-33, 1.30E-08, 1.20E-15, 2.20E-09, 2.50E-10, 1.80E-08, 2.50E-08, 1.00E-09, 4.50E-08, 8.00E-08, 2.20E-16, 2.60E-08, 1.70E-10, 3.70E-08, 3.50E-13, 1.40E-08, 5.70E-10, 7.50E-10, 1.30E-09, 5.10E-08, 7.00E-10, 1.20E-08, 6.80E-08, 3.90E-09, 4.90E-11, 2.90E-11, 1.70E-09, 8.00E-10, 2.70E-10, 2.30E-08, 6.20E-09, 1.50E-13, 1.30E-11, 3.10E-09, 1.60E-12, 2.40E-08, 6.10E-13, 3.30E-08, 3.50E-08, 3.20E-10, 4.20E-08, 2.00E-09, 2.40E-16, 6.10E-09, 4.30E-08, 6.30E-08, 1.90E-23, 1.00E-11, 8.90E-09, 9.10E-09, 3.30E-14, 5.70E-09, 4.30E-09, 1.90E-16, 2.40E-12, 7.60E-08, 4.60E-08, 6.40E-12, 2.80E-20, 2.30E-11, 1.60E-11, 5.90E-13, 4.10E-08, 1.30E-12, 4.80E-08, 8.30E-11, 1.50E-09, 1.40E-08, 2.10E-09, 5.60E-11, 9.00E-08, 3.80E-08, 5.90E-08, 1.10E-13, 6.00E-46, 3.40E-09, 1.10E-08, 2.20E-12, 1.70E-20)
```

Assuming they are all independent and there is no winner's curse the variance explained in BMI variance by these `r length(pvals)` hits is:

```{r}
tvals <- qt(pvals, 10000, low=F)
r2 <- tvals^2 / (tvals^2 + 10000)
sum(r2)
```

$R^2$ is of course constrained to 1 assuming that all the influence on BMI is through DNAm sites, which is unquestionably false. If these associations were not causal then there would be no such constraint, 

We can back calculate how much variance would have been explained for different sample sizes below 10k. Are we plateauing in our variance explained?

```{r}
library(ggplot2)
library(dplyr)

changing_n <- function(pval)
{
	tvals <- qt(pvals, 10000, low=F)
	tab <- tibble(
		n=seq(100, 10000, by=100),
		r2=NA,nsig=NA
	)
	r2 <- tvals^2 / (tvals^2 + 10000)
	for(i in 1:nrow(tab))
	{
		newt <- r2 * (tab$n[i]-1)/ (1-r2)
		newp <- pt(sqrt(newt), tab$n[i], low=F)
		tab$nsig[i] <- sum(newp < 5e-8)
		tab$r2[i] <- sum(r2[newp<5e-8])
	}
	return(tab)
}
tab <- changing_n(pvals)
ggplot(tab, aes(n, r2)) +
geom_point()
```

Nope.


